[C10D] Remove watchKey functionality from the Store.

kumpera opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105674
* #105672
* #105184
* #105164
* #105163
* __->__ #105014

The feature was never fully finished and never got any adoption but
TCPStore pays the cost of twice the number of tcp connections anyway.

While the cost of all those idle connections is minimal is doesn't come for free:

- It increases the likelyhood of a connection refused failure during the initialization stampede.
- TCPStore uses poll for checking for socket availability which scales linearly on the number of sockets regardless of their status.

kumpera committed:
[C10D] Remove watchKey functionality from the Store.

The feature was never fully finished and never got any adoption but
TCPStore pays the cost of twice the number of tcp connections anyway.

While the cost of all those idle connections is minimal is doesn't come for free:

- It increases the likelyhood of a connection refused failure during the initialization stampede.
- TCPStore uses poll for checking for socket availability which scales linearly on the number of sockets regardless of their status.

[ghstack-poisoned]

H-Huang commented on test/cpp/c10d/TCPStoreTest.cpp:
```diff
@@ -253,94 +169,6 @@ TEST(TCPStoreTest, testWatchKeyCallbackWithPrefix) {
   testWatchKeyCallback("testPrefix");
```
This test and the test above also need to be removed.

kumpera replied:
fixed

kumpera committed:
Update on "[C10D] Remove watchKey functionality from the Store."


The feature was never fully finished and never got any adoption but
TCPStore pays the cost of twice the number of tcp connections anyway.

While the cost of all those idle connections is minimal is doesn't come for free:

- It increases the likelyhood of a connection refused failure during the initialization stampede.
- TCPStore uses poll for checking for socket availability which scales linearly on the number of sockets regardless of their status.

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Remove watchKey functionality from the Store."


The feature was never fully finished and never got any adoption but
TCPStore pays the cost of twice the number of tcp connections anyway.

While the cost of all those idle connections is minimal is doesn't come for free:

- It increases the likelyhood of a connection refused failure during the initialization stampede.
- TCPStore uses poll for checking for socket availability which scales linearly on the number of sockets regardless of their status.

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Remove watchKey functionality from the Store."


The feature was never fully finished and never got any adoption but
TCPStore pays the cost of twice the number of tcp connections anyway.

While the cost of all those idle connections is minimal is doesn't come for free:

- It increases the likelyhood of a connection refused failure during the initialization stampede.
- TCPStore uses poll for checking for socket availability which scales linearly on the number of sockets regardless of their status.

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Remove watchKey functionality from the Store."


The feature was never fully finished and never got any adoption but
TCPStore pays the cost of twice the number of tcp connections anyway.

While the cost of all those idle connections is minimal is doesn't come for free:

- It increases the likelyhood of a connection refused failure during the initialization stampede.
- TCPStore uses poll for checking for socket availability which scales linearly on the number of sockets regardless of their status.

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Remove watchKey functionality from the Store."


The feature was never fully finished and never got any adoption but
TCPStore pays the cost of twice the number of tcp connections anyway.

While the cost of all those idle connections is minimal is doesn't come for free:

- It increases the likelyhood of a connection refused failure during the initialization stampede.
- TCPStore uses poll for checking for socket availability which scales linearly on the number of sockets regardless of their status.

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Remove watchKey functionality from the Store."


The feature was never fully finished and never got any adoption but
TCPStore pays the cost of twice the number of tcp connections anyway.

While the cost of all those idle connections is minimal is doesn't come for free:

- It increases the likelyhood of a connection refused failure during the initialization stampede.
- TCPStore uses poll for checking for socket availability which scales linearly on the number of sockets regardless of their status.

[ghstack-poisoned]

fduwjj reviewed (approved):
Thanks for cleaning this up. Does this mean we previously had twice as much as sockets? That's a lot.

kumpera committed:
Update on "[C10D] Remove watchKey functionality from the Store."


The feature was never fully finished and never got any adoption but
TCPStore pays the cost of twice the number of tcp connections anyway.

While the cost of all those idle connections is minimal is doesn't come for free:

- It increases the likelyhood of a connection refused failure during the initialization stampede.
- TCPStore uses poll for checking for socket availability which scales linearly on the number of sockets regardless of their status.

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Remove watchKey functionality from the Store."


The feature was never fully finished and never got any adoption but
TCPStore pays the cost of twice the number of tcp connections anyway.

While the cost of all those idle connections is minimal is doesn't come for free:

- It increases the likelyhood of a connection refused failure during the initialization stampede.
- TCPStore uses poll for checking for socket availability which scales linearly on the number of sockets regardless of their status.

[ghstack-poisoned]

kumpera commented:
@pytorchmergebot  merge

kumpera committed:
Update on "[C10D] Remove watchKey functionality from the Store."


The feature was never fully finished and never got any adoption but
TCPStore pays the cost of twice the number of tcp connections anyway.

While the cost of all those idle connections is minimal is doesn't come for free:

- It increases the likelyhood of a connection refused failure during the initialization stampede.
- TCPStore uses poll for checking for socket availability which scales linearly on the number of sockets regardless of their status.

[ghstack-poisoned]

kumpera commented:
@pytorchmergebot merge -f "failures are unrelated"
